diagnostics/metrics.rs

```diff
@@ -0,0 +1,729 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::{collections::HashMap, fmt, path::PathBuf, time::Instant};
+
+use chrono::Utc;
+use resource::constants::diagnostics::{REPORT_INTERVAL, UNKNOWN_STR};
+use serde_json::{json, map::Map as JSONMap, Value as JSONValue};
+use sysinfo::{Disks, MemoryRefreshKind, RefreshKind, System};
+
+use crate::{version::JSON_API_VERSION, DatabaseHash, DatabaseHashOpt};
+
+#[derive(Debug)]
+pub(crate) struct ServerProperties {
+    json_api_version: u64,
+    deployment_id: String,
+    server_id: String,
+    distribution: String,
+    is_reporting_enabled: bool,
+}
+
+impl ServerProperties {
+    pub(crate) fn new(
+        deployment_id: String,
+        server_id: String,
+        distribution: String,
+        is_reporting_enabled: bool,
+    ) -> ServerProperties {
+        Self { json_api_version: JSON_API_VERSION, deployment_id, server_id, distribution, is_reporting_enabled }
+    }
+
+    pub fn deployment_id(&self) -> &str {
+        &self.deployment_id
+    }
+
+    pub fn server_id(&self) -> &str {
+        &self.server_id
+    }
+
+    pub fn is_reporting_enabled(&self) -> bool {
+        self.is_reporting_enabled
+    }
+
+    pub fn to_reporting_json(&self) -> JSONValue {
```

Comment:
I was thinking about having trait(s), but it ended up being quite messy, so I gave up and implemented these methods in a straightforward way. No "overrefactoring".